src/cbc_sdk/platform/policy_ruleconfigs.py

```diff
@@ -272,3 +299,316 @@ def set_assignment_mode(self, mode):
+            if self._parent:
+                self._parent._mark_changed()
+
+        def _flatten(self):
+            """
+            Turns this rule group into a dict for transferral to the server.
+
+            Returns:
+                dict: The information defining the rule group and its constituent rules.
+            """
+            rc = copy.deepcopy(self._info)
+            rc['rules'] = [rule._flatten() for rule in self._rules]
+            return rc
+
+        @property
+        def rules_(self):
+            """
+            Returns a list of the firewall rules within this rule group.
+
+            Returns:
+                list(HostBasedFirewallRuleConfig.FirewallRule): List of contained firewall rules.
+            """
+            return self._rules
+
+        def append_rule(self, rule):
+            """
+            Appends a new rule to this rule group.
+
+            Args:
+                rule (HostBasedFirewallRuleConfig.FirewallRule): The new rule.
+            """
+            rule._parent = self._parent
+            self._rules.append(rule)
+            if self._parent:
+                self._parent._mark_changed()
+
+        def remove(self):
+            """Removes this rule group from the rule configuration."""
+            if self._parent and self in self._parent.rule_groups:
+                self._parent.rule_groups.remove(self)
+                self._parent._mark_changed()
+                self._parent = None
+
+    class FirewallRule(MutableBaseModel):
+        """Represents a single firewall rule."""
+        swagger_meta_file = "platform/models/firewall_rule.yaml"
+
+        def __init__(self, cb, initial_data):
+            """
+            Initialize the FirewallRule object.
+
+            Args:
+                cb (BaseAPI): Reference to API object used to communicate with the server.
+                initial_data (dict): Initial data used to populate the firewall rule.
+            """
+            super(HostBasedFirewallRuleConfig.FirewallRule, self).__init__(cb, None, initial_data, False, True)
+            self._parent = None
+
+        def _field_updated(self, attrname):
+            """Method called whenever a field is updated."""
+            if self._parent:
+                self._parent._mark_changed()
+
+        def _flatten(self):
+            """
+            Turns this rule into a dict for transferral to the server.
+
+            Returns:
+                dict: The information defining the rule.
+            """
+            return copy.deepcopy(self._info)
+
+        def remove(self):
+            """Removes this rule from the rule group that contains it."""
+            if self._parent:
+                group_list = [group for group in self._parent.rule_groups if self in group._rules]
+                if group_list:
+                    group_list[0]._rules.remove(self)
+                    self._parent._mark_changed()
+                    self._parent = None
+
+    def _base_url(self):
```

Comment:
I missed this in the previous rule config PRs. Why are we not implmenenting this as a `urlobject` property on the class? Why does this need to be a function?

Reply:
Because that base URL always contains the policy ID of the "parent" policy for these rule configs, and the base class implementation (in `PolicyRuleConfig`) reflects that. It's set up so that _only_ the base class implementation of `_base_url` needs to get the parent policy ID value; all the subclass base URLs are just that URL with something else tacked onto the end, as you see here.  It _could_ be written to use `urlobject`, but that wouldn't be very good encapsulation because I'd have to "leak" the parent policy ID into the subclasses.

Reply:
`_parent` already exists in every RuleConfig class as it inherits the `PolicuRuleConfig` class which is why they should all have a unique `urlobject` property `/policyservice/v1/orgs/{0}/policies/{1}/rule_configs/host_based_firewall` where it can be set by using the following code
```
self.urlobject.format(self._cb.credentials.org_key, self._parent._model_unique_id)
```

Reply:
The reason why I don't like that style is because it forces us to reference `self._parent._model_unique_id` at all levels of the class hierarchy, not _just_ in the base class. However, I'm not gonna get all religious about it, and I'll do it your way.

Reply:
I felt `urlobject_single` was more appropriate to use here, and the code has been rewritten to use it and eliminate `_base_url()` entirely.  This will be in the next commit.